provider/clickhouse.go

```diff
@@ -996,20 +1047,88 @@ func (store *clickHouseOfflineStore) GetTrainingSet(id ResourceID) (TrainingSetI
 	for _, name := range columnNames {
 		features = append(features, sanitizeCH(name.Name))
 	}
-	columns := strings.Join(features[:], ", ")
-	trainingSetQry := store.query.trainingRowSelect(columns, trainingSetName)
+	columns := strings.Join(features, ", ")
+
+	return &TrainingSetPreparation{
+		TrainingSetName: trainingSetName,
+		Columns:         columns,
+	}, nil
+}
+
+func (store *clickHouseOfflineStore) GetTrainingSet(id ResourceID) (TrainingSetIterator, error) {
+	fmt.Printf("Getting Training Set: %v\n", id)
+	prep, err := store.prepareTrainingSetQuery(id)
+	if err != nil {
+		return nil, err
+	}
+	trainingSetQry := store.query.trainingRowSelect(prep.Columns, prep.TrainingSetName)
 	fmt.Printf("Training Set Query: %s\n", trainingSetQry)
 	rows, err := store.db.Query(trainingSetQry)
 	if err != nil {
 		return nil, fferr.NewResourceExecutionError(pt.ClickHouseOffline.String(), id.Name, id.Variant, fferr.ResourceType(id.Type.String()), err)
 	}
-	colTypes, err := store.getValueColumnTypes(trainingSetName)
+	colTypes, err := store.getValueColumnTypes(prep.TrainingSetName)
 	if err != nil {
 		return nil, err
 	}
 	return store.newsqlTrainingSetIterator(rows, colTypes), nil
 }
 
+func (store *clickHouseOfflineStore) GetTrainingSetTestSplit(
+	id ResourceID,
+	testSize float32,
+	shuffle bool,
+	randomState int,
+) (TrainingSetIterator, TrainingSetIterator, func() error, error) {
```

Comment:
Dont love having this many things returned, I have advice in a comment below though